pydantic/types.py

```diff
@@ -1333,12 +1334,45 @@ def encode_str(self, value: str) -> str:
 
 
 @_internal_dataclass.slots_dataclass
-class TransformSchema:
-    """An annotation that can be used to apply a transform to a core schema."""
+class GetPydanticSchema:
+    """A convenience class for creating an annotation that provides pydantic custom type hooks.
 
-    transform: Callable[[CoreSchema], CoreSchema]
+    This class lets you replace the use of a custom type defining one or more of `__get_pydantic_core_schema__`,
+    `__get_pydantic_json_schema__`, and `__prepare_pydantic_annotations__` with a single annotation.
+    """
 
-    def __get_pydantic_core_schema__(
-        self, source_type: type[Any], handler: _annotated_handlers.GetCoreSchemaHandler
-    ) -> CoreSchema:
-        return self.transform(handler(source_type))
+    get_pydantic_core_schema: Callable[[Any, _annotated_handlers.GetCoreSchemaHandler], CoreSchema] | None = None
+    get_pydantic_json_schema: Callable[[Any, _annotated_handlers.GetJsonSchemaHandler], JsonSchemaValue] | None = None
+    prepare_pydantic_annotations: Callable[[Any, tuple[Any, ...], ConfigDict], tuple[Any, Iterable[Any]]] | None = None
+
+    @staticmethod
+    def for_type(type_: Any) -> GetPydanticSchema:
```

Comment:
Can we leave this off for now? I'd like it to be requested/ used in more places before we introduce an abstraction.

Reply:
If we don't include it, I would at least consider my proposed solution above to the numpy thing a request.

In particular, I think `GetPydanticSchema.for_type(InstanceOf[np.ndarray])` is a lot easier to understand than
`GetPydanticSchema(lambda _s, h: h(InstanceOf[np.ndarray]), lambda _s, h: h(InstanceOf[np.ndarray]))`.

I'm okay waiting until I have another example or two where this utility would be useful, but I think it's advanced enough that I'll be surprised if people request it even if they would find it useful; I sort of think we'd need documentation of it before people would realize it was an option. Basically, if we internally want it a few more times I think it would be worth including.

The best argument against including it now, in my opinion, is if we want/need to change the way it interacts with `__prepare_pydantic_annotations__`, and just suggesting users use the thing I said was "hard to understand" above.

Reply:
hmm actually I think I made a mistake in the `for_type` method anyway, and you don't need to specify the `get_pydantic_json_schema` if you specify the core one. So maybe that's another good reason not to include it 😅. I'll remove it for now.